### PR TITLE
feat(redis): add logic operators $and and $or in redis

### PR DESCRIPTION
**Pull request by AnneYang720**
This PR will support logic operators $and and $or in redis according to https://github.com/jina-ai/docarray/issues/501

### CODE REVIEW

File: docarray/array/storage/redis/find.py [hunk 1]
```diff
@@ -32,10 +43,14 @@ def _find_similar_vectors(
         **kwargs,
     ):
 
-        query_str = self._build_query_str(filter) if filter else "*"
+        if filter:
+            nodes = _build_query_nodes(filter)
+            query_str = intersect(*nodes).to_string()
+        else:
+            query_str = "*"
```
Comment (bwanglzu) [inline docarray/array/storage/redis/find.py:50]:
```suggestion
            query_str = '*'
```

File: docarray/array/storage/redis/find.py [hunk 2]
```diff
@@ -92,36 +108,59 @@ def _filter(
...
+        if operator == '$gt':
+            query_dict[key] = gt(value)
+        elif operator == '$gte':
+            query_dict[key] = ge(value)
+        elif operator == '$lt':
+            query_dict[key] = lt(value)
+        elif operator == '$lte':
+            query_dict[key] = le(value)
+        elif operator == '$eq':
+            if isinstance(value, (bool)):
```
Comment (bwanglzu) [inline docarray/array/storage/redis/find.py:136]: what is `(bool)`?

File: docarray/array/storage/redis/find.py [hunk 3]
```diff
@@ -92,36 +108,59 @@ def _filter(
...
+            )
+        node = IntersectNode(**query_dict)
+
+    return node
+
+
+def _build_query_nodes(filter):
+    nodes = []
+    for k, v in filter.items():
+        if k == "$and":
```
Comment (bwanglzu) [inline docarray/array/storage/redis/find.py:154]: single quote

File: tests/unit/array/mixins/test_find.py [hunk 4]
```diff
@@ -543,40 +548,79 @@ def test_redis_category_filter(start_storage):
...
-    assert all([(r.tags['isfake'] == False) for r in results])
+    assert all(
+        [
+            (
+                (r.tags['price'] >= 8 and r.tags['category'] != 'Shoes')
+                or r.tags['size'] == 3
+            )
+            for r in results
+        ]
+    )
```
Comment (bwanglzu) [inline tests/unit/array/mixins/test_find.py:623]: your `filters1` to `filters4` can be parameters and your expected output can be parameters to simplify your test

**Review by bwanglzu — CHANGES_REQUESTED** [review]
some minor comments, please always use single quote.

File: docarray/array/storage/redis/find.py [hunk 5]
```diff
@@ -92,36 +108,59 @@ def _filter(
...
+                query_dict[key] = equal(value)
+            else:
+                query_dict[key] = value
+        else:
+            raise ValueError(
+                f'Expecting filter operator one of $gt, $gte, $lt, $lte, $eq, $ne, $and OR $or, got {operator} instead'
+            )
+        node = IntersectNode(**query_dict)
+
+    return node
```
Comment (bwanglzu) [inline docarray/array/storage/redis/find.py:148]: i modified it a bit for simplicity:

```python

def _build_query_node(key, condition):
    operator = list(condition.keys())[0]
    value = condition[operator]

    query_dict = {}

    if operator in ['$ne', 'eq']:
        if isinstance(value, bool):
            query_dict[key] = equal(int(value))
        elif isinstance(value, (int, float)):
            query_dict[key] = equal(value)
        else:
            query_dict[key] = value
    elif operator == '$gt':
        query_dict[key] = gt(value)
    elif operator == '$gte':
        query_dict[key] = ge(value)
    elif operator == '$lt':
        query_dict[key] = lt(value)
    elif operator == '$lte':
        query_dict[key] = le(value)
    else:
        raise ...

    if operator == '$ne':
        return DistjunctUnion(**query_dict)
    return IntersectNode(**query_dict)
```

**Review by bwanglzu — CHANGES_REQUESTED** [review]
minor comments

**Review by bwanglzu — APPROVED** [review]
LGTM!

**Review by JoanFM — CHANGES_REQUESTED** [review]
Wait for merging PR where column types is set to Dict

File: docs/advanced/document-store/redis.md [hunk 6]
```diff
@@ -165,58 +171,119 @@ da.extend(
...
 ):
-    print(f'\tembedding={embedding},\t price={price},\t color={color}')
+    print(f'\tembedding={embedding},\t color={color},\t stock={stock}')
 ```
 
 Consider the case where you want the nearest vectors to the embedding `[8.,  8.,  8.]`, with the restriction that
-prices and colors must pass a filter. For example, let's consider that retrieved Documents must have a `price` value lower than or equal to `max_price` and have `color` equal to `color`. We can encode this information in Redis using `{'price': {'$lte': max_price}, 'color': {'$eq': color}}`.
+prices, colors and stock must pass a filter. For example, let's consider that retrieved Documents must have a `price` value lower than or equal to `max_price`, have `color` equal to `color` and have `stock` equal to `True`. We can encode this information in Redis using 
+
+```JSON
```
[the remark this answers is not in the document]
Comment (AnneYang720) [reply docs/advanced/document-store/redis.md:190]: okay

**Review by alaeddine-13 — APPROVED** [review]
LGTM, well done !

**Review by JoanFM — APPROVED** [review]

**Pull request merged**